src/order/locally_finite.lean

```diff
@@ -0,0 +1,516 @@
+* `multiset.Ici`: Closed-infinite interval as a multiset.
+* `multiset.Ioi`: Open-infinite interval as a multiset.
+
+When it's also an `order_bot`,
+* `finset.Iic`: Infinite-open interval as a finset.
+* `finset.Iio`: Infinite-closed interval as a finset.
+* `multiset.Iic`: Infinite-open interval as a multiset.
+* `multiset.Iio`: Infinite-closed interval as a multiset.
+
+## Instances
+
+A `locally_finite_order` instance can be built
+* for a subtype of a locally finite order. See `subtype.locally_finite_order`.
+* for the product of two locally finite orders. See `prod.locally_finite_order`.
+* for any fintype (but it is noncomputable). See `fintype.locally_finite_order`.
+* from a definition of `finset.Icc` alone. See `locally_finite_order.of_Icc`.
+* by pulling back `locally_finite_order β` through an order embedding `f : α →o β`. See
+  `order_embedding.locally_finite_order`.
+
+Instances for concrete types are proved in their respective files:
+* `ℕ` is in `data.nat.intervals`
+* `ℤ` is in `data.int.intervals`
+* `ℕ+` is in `data.pnat.intervals`
+* `fin n` is in `data.fin.intervals`
+Along, you will find lemmas about the cardinality of those finite intervals.
+
+## TODO
+
+`multiset.Ico` hasn't been generalized yet. All of `data.multiset.intervals` should be generalized.
+
+Provide the `locally_finite_order` instance for `lex α β`.
+
+From `linear_order α`, `no_top_order α`, `locally_finite_order α`, we can also define an
+order isomorphism `α ≃ ℕ` or `α ≃ ℤ`, depending on whether we have `order_bot α` or
+`no_bot_order α` and `nonempty α`. When `order_bot α`, we can match `a : α` to `(Iio a).card`.
+
+Once we have the `succ_order` typeclass (any non-top element has a least greater element), we
+can provide `succ_order α` from `linear_order α` and `locally_finite_order α` using
+
+```lean
+lemma exists_min_greater [linear_order α] [locally_finite_order α] {x ub : α} (hx : x < ub) :
+  ∃ lub, x < lub ∧ ∀ y, x < y → lub ≤ y :=
+begin -- very non golfed
+  have h : (finset.Ioc x ub).nonempty := ⟨ub, finset.mem_Ioc_iff.2 ⟨hx, le_rfl⟩⟩,
+  use finset.min' (finset.Ioc x ub) h,
+  split,
+  { have := finset.min'_mem _ h,
+    simp * at * },
+  rintro y hxy,
+  obtain hy | hy := le_total y ub,
+  apply finset.min'_le,
+  simp * at *,
+  exact (finset.min'_le _ _ (finset.mem_Ioc_iff.2 ⟨hx, le_rfl⟩)).trans hy,
+end
+```
+Note that the converse is not true. Consider `{-2^z | z : ℤ} ∪ {2^z | z : ℤ}`. Any element has a
+successor (and actually a predecessor as well), so it is a `succ_order`, but it's not locally finite
+as `Icc (-1) 1` is infinite.
+-/
+
+open finset
+
+/-- A locally finite order is an order where bounded intervals are finite. When you don't care too
+much about definitional equality, you can use `locally_finite_order.of_Icc` or
+`locally_finite_order.of_finite_Icc` to build a locally finite order from just `finset.Icc`. -/
+@[ext] class locally_finite_order (α : Type*) [preorder α] :=
```

Comment:
I don't think `ext` makes sense, as this is subsingleton

Reply:
I needed it precisely to use `ext` in proof of the subsingleton instance. How do I make do manually?

Reply:
Just use `cases` in the proof?

Reply:
Oh, duh. Of course